docs/content/configuration/config-file.md
# TOML Configuration File

**`SWS`** can be configured using a [TOML](https://toml.io/en/) file to adjust the general server features as well as other advanced ones.

It's disabled by default and can be enabled by passing a _string file path_ via the `-w, --config-file` option or its equivalent [SERVER_CONFIG_FILE](./../configuration/environment-variables.md#server_config_file) env.

!!! info "The default config file path is checked at startup time"

    If using the default config file path (`./sws.toml`), SWS will attempt to load it at startup time.
    If it is not found or can not be loaded then SWS will continue using the server defaults.

## TOML File (Manifest)

Below is just an example showing all features with their default values.

```toml
[general]

#### Address & Root dir
host = "::"
port = 80
root = "./public"

#### Logging
log-level = "error"
log-format = "json"

#### Cache Control headers
cache-control-headers = true

#### Auto Compression
compression = true
compression-level = "default"

#### Error pages
# Note: If a relative path is used then it will be resolved under the root directory.
page404 = "./404.html"
page50x = "./50x.html"

#### HTTP/2 + TLS
http2 = false
http2-tls-cert = ""
http2-tls-key = ""
https-redirect = false
https-redirect-host = "localhost"
https-redirect-from-port = 80
https-redirect-from-hosts = "localhost"

#### CORS & Security headers
# security-headers = true
# cors-allow-origins = ""

#### Directory listing
directory-listing = false

#### Directory listing sorting code
directory-listing-order = 1

#### Directory listing content format
directory-listing-format = "html"

#### Directory listing download format
directory-listing-download = []

#### Basic Authentication
# basic-auth = ""

#### File descriptor binding
# fd = ""

#### Worker threads
threads-multiplier = 1

#### Grace period after a graceful shutdown
grace-period = 0

#### Page fallback for 404s
# page-fallback = ""

#### Log request Remote Address if available
log-remote-address = false

#### Log real IP from X-Forwarded-For header if available
log-forwarded-for = false

#### IPs to accept the X-Forwarded-For header from. Empty means all
trusted-proxies = []

#### Redirect to trailing slash in the requested directory uri
redirect-trailing-slash = true

#### Check for existing pre-compressed files
compression-static = true

#### Health-check endpoint (GET or HEAD `/health`)
health = false

#### Markdown content negotiation
accept-markdown = false

#### Default charset for text responses (enabled by default)
# text-charset = false

#### List of index files
# index-files = "index.html, index.htm"
#### Maintenance Mode

maintenance-mode = false
# maintenance-mode-status = 503
# maintenance-mode-file = "./maintenance.html"

### Windows Only

#### Run the web server as a Windows Service
# windows-service = false


[advanced]

#### HTTP Headers customization (examples only)

#### a. Oneline version
# [[advanced.headers]]
# source = "**/*.{js,css}"
# headers = { Access-Control-Allow-Origin = "*" }

#### b. Multiline version
# [[advanced.headers]]
# source = "/index.html"
# [advanced.headers.headers]
# Cache-Control = "public, max-age=36000"
# Content-Security-Policy = "frame-ancestors 'self'"
# Strict-Transport-Security = "max-age=63072000; includeSubDomains; preload"

#### c. Multiline version with explicit key (dotted)
# [[advanced.headers]]
# source = "**/*.{jpg,jpeg,png,ico,gif}"
# headers.Strict-Transport-Security = "max-age=63072000; includeSubDomains; preload"


### URL Redirects (examples only)

# [[advanced.redirects]]
# source = "**/*.{jpg,jpeg}"
# destination = "/images/generic1.png"
# kind = 301

# [[advanced.redirects]]
# source = "/index.html"
# destination = "https://static-web-server.net"
# kind = 302

### URL Rewrites (examples only)

# [[advanced.rewrites]]
# source = "**/*.{png,ico,gif}"
# destination = "/assets/favicon.ico"
## Optional redirection
# redirect = 301

# [[advanced.rewrites]]
# source = "**/*.{jpg,jpeg}"
# destination = "/images/sws.png"

### Virtual Hosting

# [[advanced.virtual-hosts]]
## But if the "Host" header matches this...
# host = "sales.example.com"
## ...then files will be served from here instead
# root = "/var/sales/html"

# [[advanced.virtual-hosts]]
# host = "blog.example.com"
# root = "/var/blog/html"

# [advanced.memory-cache]
# capacity = 100
# ttl = 1800      # 30 minutes
# tti = 300       # 5 minutes
# max-file-size = 8192  # 8 MiB
```

### General options

The TOML `[general]` section allows adjusting the current options available via the CLI/ENV ones.

So they are equivalent to each other **except** for the `-w, --config-file` option which is omitted and can not be used for obvious reasons.

!!! info "Config file-based features are optional"

    All server feature options via the configuration file are optional and can be omitted as needed.

### Advanced options

The TOML `[advanced]` section is intended for more complex features.

For example [Custom HTTP Headers](../features/custom-http-headers.md), [Custom URL Redirects](../features/url-redirects.md), [URL Rewrites](../features/url-rewrites.md), [Virtual Hosting](../features/virtual-hosting.md), or [In-Memory Cache](../features/memory-cache.md).

### Precedence

Whatever config file-based feature option will take precedence over its CLI or ENV equivalent.

## Usage

The following command runs the server using a specific `sws.toml` file.

```sh
static-web-server -w sws.toml
```


docs/configuration/environment-variables.md
# Environment Variables

Server can be configured via the following environment variables arguments.

!!! tip "Remember"
    - Environment variables are equivalent to their command-line arguments.
    - [Command line arguments](./environment-variables.md) take precedence over their environment variables.

### SERVER_HOST
The address of the host (E.g 127.0.0.1). Default `[::]`.

### SERVER_PORT
The port of the host. Default `80`.

### SERVER_LISTEN_FD
Optional file descriptor number (e.g. `0`) to inherit an already-opened TCP listener on (instead of using `SERVER_HOST` and/or `SERVER_PORT`).

### SERVER_ROOT
Relative or absolute root directory path of static files. Default `./public`.

### SERVER_LOG_LEVEL
Specify a logging level in lower case. Possible values are `error`, `warn`, `info`, `debug` or `trace`. Default `error`.

### SERVER_ERROR_PAGE_404
HTML file path for 404 errors. If path is not specified or simply don't exists then server will use a generic HTML error message. Default `./public/404.html`.

### SERVER_ERROR_PAGE_50X
HTML file path for 50x errors. If path is not specified or simply don't exists then server will use a generic HTML error message. Default `./public/50x.html`

### SERVER_THREADS_MULTIPLIER
Number of worker threads multiplier that'll be multiplied by the number of system CPUs using the formula: `worker threads = number of CPUs * n` where `n` is the value that changes here. When multiplier value is 0 or 1 then the `number of CPUs` is used. Number of worker threads result should be a number between 1 and 32,768 though it is advised to keep this value on the smaller side. Default one thread per core.

### SERVER_HTTP2_TLS
Enable HTTP/2 with TLS support. Make sure also to adjust current server port. Default `false` (disabled).

### SERVER_HTTP2_TLS_CERT
Specify the file path to read the certificate. Default empty.

### SERVER_HTTP2_TLS_KEY
Specify the file path to read the private key. Default empty.

### SERVER_CORS_ALLOW_ORIGINS
Specify a optional CORS list of allowed origin hosts separated by comas. Host ports or protocols aren't being checked. Use an asterisk (*) to allow any host. Default empty (disabled).

### SERVER_COMPRESSION
Gzip, Deflate or Brotli compression on demand determined by the *Accept-Encoding* header and applied to text-based web file types only. See [ad-hoc mime-type list](https://github.com/joseluisq/static-web-server/blob/master/src/compression.rs#L20). Default `true` (enabled).

### SERVER_DIRECTORY_LISTING
Enable directory listing for all requests ending with the slash character (‘/’). Default `false` (disabled).

### SERVER_SECURITY_HEADERS
Enable security headers by default when HTTP/2 feature is activated. Headers included: `Strict-Transport-Security: max-age=63072000; includeSubDomains; preload` (2 years max-age), `X-Frame-Options: DENY`, `X-XSS-Protection: 1; mode=block` and `Content-Security-Policy: frame-ancestors 'self'`. Default `false` (disabled).

### SERVER_CACHE_CONTROL_HEADERS
Enable cache control headers for incoming requests based on a set of file types. The file type list can be found on [`src/control_headers.rs`](./src/control_headers.rs) file. Default `true` (enabled).

### SERVER_BASIC_AUTH
It provides [The "Basic" HTTP Authentication Scheme](https://datatracker.ietf.org/doc/html/rfc7617) using credentials as `user-id:password` pairs, encoded using `Base64`. Password must be encoded using the [BCrypt](https://en.wikipedia.org/wiki/Bcrypt) password-hashing function. Default empty (disabled).


CHANGELOG.md
# Static Web Server v2 - Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v2.42.0 - 2026-03-27

This new `v2.42.0` brings bug fixes, new features and improvements. Fix a memory increase regression introduced in `v2.40.0`. Support for HTTP Prometheus metrics, Local-time for logs by default, support for the POSIX `TZ` environment variable to configure logs's timezone as well as other minor improvements.

**Fixes**

- [0b128b9][0b128b9] Update dependencies and bump up Rust to `1.88.0`. PR [#641][641] by [@joseluisq][joseluisq]
- [4124999][4124999] Regression: Memory increase for Linux Musl statically-linked binaries introduced in `v2.40.0`. More details in PR [#640][640] by [@joseluisq][joseluisq]
- [8451cf7][8451cf7] `--disable-symlinks` option does not work properly if a path contains intermediate symlink components. PR [#639][639] by [@joseluisq][joseluisq]

[4124999]: https://github.com/static-web-server/static-web-server/commit/4124999
[8451cf7]: https://github.com/static-web-server/static-web-server/commit/8451cf7

[640]: https://github.com/static-web-server/static-web-server/pull/640
[641]: https://github.com/static-web-server/static-web-server/pull/641
[639]: https://github.com/static-web-server/static-web-server/pull/639

**Features**

- [b798f68][b798f68] Local-time support for logs by default to honor user's system time. It also supports the POSIX `TZ` environment variable to update the logs's time zone on demand. PR [#632][632] by [@joseluisq][joseluisq]. See [docs](https://static-web-server.net/features/logging/#log-date-timestamp).
- [2d50f88][2d50f88] Stabilize Prometheus metrics feature via `--metrics` option. PR [#635][635] by [@chrissnell][chrissnell]. See [docs](https://static-web-server.net/features/metrics)
- [a4213e5][a4213e5] HTTP-level Prometheus metrics to the metrics endpoint. PR [#631][631] by [@chrissnell][chrissnell]
- [6ada726][6ada726] Grafana dashboard example for Prometheus metrics feature. PR [#636][636] by [@chrissnell][chrissnell]. See [example](https://github.com/static-web-server/static-web-server/tree/master/contrib/grafana).

[0b128b9]: https://github.com/static-web-server/static-web-server/commit/0b128b9
[2d50f88]: https://github.com/static-web-server/static-web-server/commit/2d50f88
[b798f68]: https://github.com/static-web-server/static-web-server/commit/b798f68
[6ada726]: https://github.com/static-web-server/static-web-server/commit/6ada726
[a4213e5]: https://github.com/static-web-server/static-web-server/commit/a4213e5

[635]: https://github.com/static-web-server/static-web-server/pull/635
[632]: https://github.com/static-web-server/static-web-server/pull/632
[636]: https://github.com/static-web-server/static-web-server/pull/636
[631]: https://github.com/static-web-server/static-web-server/pull/631

**Docs**

- [0a8c8ca][0a8c8ca] Metrics feature documentation page. PR [#633][633] by [@chrissnell][chrissnell] See [docs](https://static-web-server.net/features/metrics)
- [b6856ea][b6856ea] Local-time support for logs. PR [#638][638] by [@joseluisq][joseluisq]

[0a8c8ca]: https://github.com/static-web-server/static-web-server/commit/0a8c8ca
[b6856ea]: https://github.com/static-web-server/static-web-server/commit/b6856ea

[633]: https://github.com/static-web-server/static-web-server/pull/633
[638]: https://github.com/static-web-server/static-web-server/pull/638
[627]: https://github.com/static-web-server/static-web-server/pull/627

[chrissnell]: https://github.com/chrissnell

For more details see the [v2.42.0 milestone][v2420-milestone] and the full changelog [v2.41.0...v2.42.0][v2420-diff].

[v2420-diff]: https://github.com/static-web-server/static-web-server/compare/v2.41.0...v2.42.0
[v2420-milestone]: https://github.com/static-web-server/static-web-server/milestone/37?closed=1

## v2.41.0 - 2026-02-20

This new `v2.41.0` release includes important bug fixes, new features, and improvements.
The custom headers, installer scripts and hidden file handling are now more reliable. Dynamic compression encodings have been enhanced with internal priority support based on modern compression algorithms. Default options now help protect hidden files and prevent risky symlink usage. The installer and Docker images have been improved and dependencies are updated for better stability. Also, the documentation features a new `showcases` page to highlight how SWS is being used in the wild.

**Security Patch**

This particular release patches a security vulnerability targeting the Basic Authentication feature.
More details about the fixed security issue will be provided in the next hours.

We encourage users to update as soon as possible.

**Fixes**

- [88422ba][88422ba] Update project dependencies. ([#620][620]) by [@joseluisq][joseluisq]
- [bc7b7cd][bc7b7cd] Docker: Update Alpine (3.22.3) & Debian (13.3) Docker images. ([#619][619],[#625][625]) by [@joseluisq][joseluisq]
- [df5fb00][df5fb00] Custom headers are not applied when `--redirect-trailing-slash` is disabled. ([#613][613]) by [@joseluisq][joseluisq]
- [1a31f00][1a31f00] Hidden root paths (e.g. `.public`) are ignored by `--ignore-hidden-files` feature. ([#606][606]) by [@mightyiam][mightyiam]
- [9fbafcf][9fbafcf] Installer script fails when used in Alpine Linux. ([#610][610]) by [@joseluisq][joseluisq]
- [c298a6d][c298a6d] CI: Remove `pull_request` trigger from `release-docker-devel` workflow. ([#608][608]) by [@joseluisq][joseluisq]
- [7bf0fd4][7bf0fd4] Timing-based username enumeration issue in 'Basic Authentication' feature. By [@naoyashiga][naoyashiga]

[88422ba]: https://github.com/static-web-server/static-web-server/commit/88422ba
[bc7b7cd]: https://github.com/static-web-server/static-web-server/commit/bc7b7cd
[c298a6d]: https://github.com/static-web-server/static-web-server/commit/c298a6d

src/config.rs
//! Server CLI Options

use std::path::PathBuf;

use structopt::StructOpt;

#[derive(Debug, StructOpt)]
#[structopt(about, author)]
pub struct Config {
    #[structopt(long, short = "a", default_value = "::", env = "SERVER_HOST")]
    /// Host address (E.g 127.0.0.1 or ::1)
    pub host: String,

    #[structopt(long, short = "p", default_value = "80", env = "SERVER_PORT")]
    /// Host port
    pub port: u16,

    #[structopt(
        long,
        short = "f",
        env = "SERVER_LISTEN_FD",
        conflicts_with_all(&["host", "port"])
    )]
    /// Instead of binding to a TCP port, accept incoming connections to an already-bound TCP
    /// socket listener on the specified file descriptor number (usually zero). Requires that the
    /// parent process (e.g. inetd, launchd, or systemd) binds an address and port on behalf of
    /// static-web-server, before arranging for the resulting file descriptor to be inherited by
    /// static-web-server. Cannot be used in conjunction with the port and host arguments. The
    /// included systemd unit file utilises this feature to increase security by allowing the
    /// static-web-server to be sandboxed more completely.
    pub fd: Option<usize>,

    #[structopt(
        long,
        short = "n",
        default_value = "1",
        env = "SERVER_THREADS_MULTIPLIER"
    )]
    /// Number of worker threads multiplier that'll be multiplied by the number of system CPUs
    /// using the formula: `worker threads = number of CPUs * n` where `n` is the value that changes here.
    /// When multiplier value is 0 or 1 then one thread per core is used.
    /// Number of worker threads result should be a number between 1 and 32,768 though it is advised to keep this value on the smaller side.
    pub threads_multiplier: usize,

    #[structopt(long, short = "d", default_value = "./public", env = "SERVER_ROOT")]
    /// Root directory path of static files.
    pub root: String,

    #[structopt(
        long,
        default_value = "./public/50x.html",
        env = "SERVER_ERROR_PAGE_50X"
    )]
    /// HTML file path for 50x errors. If the path is not specified or simply doesn't exist then the server will use a generic HTML error message.
    pub page50x: String,

    #[structopt(
        long,
        default_value = "./public/404.html",
        env = "SERVER_ERROR_PAGE_404"
    )]
    /// HTML file path for 404 errors. If the path is not specified or simply doesn't exist then the server will use a generic HTML error message.
    pub page404: String,

    #[structopt(long, default_value = "", env = "SERVER_FALLBACK_PAGE")]
    /// HTML file path that is used for GET requests when the requested path doesn't exist. The fallback page is served with a 200 status code, useful when using client routers. If the path is not specified or simply doesn't exist then this feature will not be active.
    pub page_fallback: String,

    #[structopt(long, short = "g", default_value = "error", env = "SERVER_LOG_LEVEL")]
    /// Specify a logging level in lower case. Values: error, warn, info, debug or trace
    pub log_level: String,

    #[structopt(
        long,
        short = "c",
        default_value = "",
        env = "SERVER_CORS_ALLOW_ORIGINS"
    )]
    /// Specify an optional CORS list of allowed origin hosts separated by comas. Host ports or protocols aren't being checked. Use an asterisk (*) to allow any host.
    pub cors_allow_origins: String,

    #[structopt(
        long,
        short = "j",
        default_value = "origin, content-type",
        env = "SERVER_CORS_ALLOW_HEADERS"
    )]
    /// Specify an optional CORS list of allowed headers separated by comas. Default "origin, content-type". It requires `--cors-allow-origins` to be used along with.
    pub cors_allow_headers: String,

    #[structopt(
        long,
        short = "t",
        parse(try_from_str),
        default_value = "false",
        env = "SERVER_HTTP2_TLS"
    )]
    /// Enable HTTP/2 with TLS support.
    pub http2: bool,

    #[structopt(
        long,
        required_if("http2", "true"),
        default_value = "",
        env = "SERVER_HTTP2_TLS_CERT"
    )]
    /// Specify the file path to read the certificate.
    pub http2_tls_cert: String,

    #[structopt(
        long,
        required_if("http2", "true"),
        default_value = "",
        env = "SERVER_HTTP2_TLS_KEY"
    )]
    /// Specify the file path to read the private key.
    pub http2_tls_key: String,

    #[structopt(
        long,
        short = "x",
        parse(try_from_str),
        default_value = "true",
        env = "SERVER_COMPRESSION"
    )]
    /// Gzip, Deflate or Brotli compression on demand determined by the Accept-Encoding header and applied to text-based web file types only.
    pub compression: bool,

    #[structopt(
        long,
        short = "z",
        parse(try_from_str),
        default_value = "false",
        env = "SERVER_DIRECTORY_LISTING"
    )]
    /// Enable directory listing for all requests ending with the slash character (‘/’).
    pub directory_listing: bool,

    #[structopt(
        long,
        required_if("directory_listing", "true"),
        default_value = "6",
        env = "SERVER_DIRECTORY_LISTING_ORDER"
    )]
    /// Specify a default code number to order directory listing entries per `Name`, `Last modified` or `Size` attributes (columns). Code numbers supported: 0 (Name asc), 1 (Name desc), 2 (Last modified asc), 3 (Last modified desc), 4 (Size asc), 5 (Size desc). Default 6 (unordered)
    pub directory_listing_order: u8,

    #[structopt(
        long,
        parse(try_from_str),
        required_if("http2", "true"),
        default_value_if("http2", Some("true"), "true"),
        default_value = "false",
        env = "SERVER_SECURITY_HEADERS"
    )]
    /// Enable security headers by default when HTTP/2 feature is activated.
    /// Headers included: "Strict-Transport-Security: max-age=63072000; includeSubDomains; preload" (2 years max-age),
    /// "X-Frame-Options: DENY", "X-XSS-Protection: 1; mode=block" and "Content-Security-Policy: frame-ancestors 'self'".
    pub security_headers: bool,

    #[structopt(
        long,
        short = "e",
        parse(try_from_str),
        default_value = "true",
        env = "SERVER_CACHE_CONTROL_HEADERS"
    )]
    /// Enable cache control headers for incoming requests based on a set of file types. The file type list can be found on `src/control_headers.rs` file.
    pub cache_control_headers: bool,

    /// It provides The "Basic" HTTP Authentication scheme using credentials as "user-id:password" pairs. Password must be encoded using the "BCrypt" password-hashing function.
    #[structopt(long, default_value = "", env = "SERVER_BASIC_AUTH")]
    pub basic_auth: String,

    #[structopt(long, short = "q", default_value = "0", env = "SERVER_GRACE_PERIOD")]
    /// Defines a grace period in seconds after a `SIGTERM` signal is caught which will delay the server before to shut it down gracefully. The maximum value is 255 seconds.
    pub grace_period: u8,

    #[structopt(long, short = "w", env = "SERVER_CONFIG_FILE")]
    /// Server TOML configuration file path.
    pub config_file: Option<PathBuf>,
}
